### PR TITLE
feat: [CDS-48385]: ASG Canary deploy step

### PR DESCRIPTION
**Pull request by vikrant-harness**
### Summary

https://user-images.githubusercontent.com/106532291/208667863-acef7afd-e99d-4a7b-bc3d-7d4e36dfdfa0.mov

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by vikrant-harness** [comment]
retrigger titlecheck

**Comment by vikrant-harness** [comment]
retrigger titlecheck

**Review by AshwinBhatkal-Harness — APPROVED** [review]

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+        allowableTypes={allowableTypes}
+        onChange={values => onChange?.(this.processFormData(values))}
+        ref={formikRef}
+        readonly={readonly}
+      />
+    )
+  }
+
+  protected type = StepType.AsgCanaryDeploy
+  protected stepName = 'Asg Canary Deploy'
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:297]: Can you please explain here how a class member be initialised from the YAML ?

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+import { getInstanceDropdownSchema } from '@common/components/InstanceDropdownField/InstanceDropdownField'
+import { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
+import { PipelineStep } from '@pipeline/components/PipelineSteps/PipelineStep'
+import { useVariablesExpression } from '@pipeline/components/PipelineStudio/PiplineHooks/useVariablesExpression'
+import type { StringsMap } from 'stringTypes'
+import { isExecutionTimeFieldDisabled } from '@pipeline/utils/runPipelineUtils'
+import stepCss from '@pipeline/components/PipelineSteps/Steps/Steps.module.scss'
+import pipelineVariablesCss from '@pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.module.scss'
+
+export interface AsgCanaryDeployData extends StepElementConfig {
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:39]: Done

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+  spec: any
+  identifier: string
+}
+
+export interface AsgCanaryDeployVariableStepProps {
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:44]: Done

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+import { StepType } from '@pipeline/components/PipelineSteps/PipelineStepInterface'
+import { PipelineStep } from '@pipeline/components/PipelineSteps/PipelineStep'
+import { useVariablesExpression } from '@pipeline/components/PipelineStudio/PiplineHooks/useVariablesExpression'
+import type { StringsMap } from 'stringTypes'
+import { isExecutionTimeFieldDisabled } from '@pipeline/utils/runPipelineUtils'
+import stepCss from '@pipeline/components/PipelineSteps/Steps/Steps.module.scss'
+import pipelineVariablesCss from '@pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.module.scss'
+
+export interface AsgCanaryDeployData extends StepElementConfig {
+  spec: any
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:40]: There is an ongoing issue with the types for this the BE will fix in their subsequent pr so i ll update accordingly

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+  props: AsgCanaryDeployProps,
+  formikRef: StepFormikFowardRef<AsgCanaryDeployData>
+): React.ReactElement {
+  const { initialValues, onUpdate, isNewStep = true, readonly, allowableTypes, stepViewType, onChange } = props
+  const { getString } = useStrings()
+  const { expressions } = useVariablesExpression()
+  return (
+    <>
+      <Formik<AsgCanaryDeployData>
+        onSubmit={(values: AsgCanaryDeployData) => {
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:74]: The interface is used for studio component and input set component , which is required in one and not in other hence the optional check

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+function AsgCanaryDeployWidget(
+  props: AsgCanaryDeployProps,
+  formikRef: StepFormikFowardRef<AsgCanaryDeployData>
+): React.ReactElement {
+  const { initialValues, onUpdate, isNewStep = true, readonly, allowableTypes, stepViewType, onChange } = props
+  const { getString } = useStrings()
+  const { expressions } = useVariablesExpression()
+  return (
+    <>
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:72]: Done

File: src/services/cd-ng/index.tsx [hunk 7]
```diff
@@ -19083,8 +19089,8 @@ export const updateApiKeyPromise = (
 export interface GetACRRegistriesBySubscriptionQueryParams {
   connectorRef: string
   accountIdentifier: string
-  orgIdentifier: string
-  projectIdentifier: string
+  orgIdentifier?: string
```
Comment (pankajsingh-harness) [inline src/services/cd-ng/index.tsx:19092]: Could you please check with the BE team why this has been made optional?

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+                  label={getString('common.instanceLabel')}
+                  readonly={readonly}
+                  expressions={expressions}
+                  allowableTypes={allowableTypes}
+                  disabledType
+                />
+                {(getMultiTypeFromValue(values?.spec?.instanceSelection?.spec?.count) === MultiTypeInputType.RUNTIME ||
+                  getMultiTypeFromValue(values?.spec?.instanceSelection?.spec?.percentage) ===
+                    MultiTypeInputType.RUNTIME) && (
+                  <ConfigureOptions
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:151]: Done

File: src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+                    isIdentifierEditable={isNewStep}
+                    inputGroupProps={{
+                      placeholder: getString('pipeline.stepNamePlaceholder'),
+                      disabled: readonly
+                    }}
+                  />
+                </div>
+              )}
+              <div className={cx(stepCss.formGroup, stepCss.sm)}>
+                <FormMultiTypeDurationField
```
[the remark this answers is not in the document]
Comment (vikrant-harness) [reply src/modules/75-cd/components/PipelineSteps/AsgCanaryDeploy/AsgCanaryDeployStep.tsx:110]: Done

File: src/services/cd-ng/index.tsx (hunk 7 above)
Comment (vikrant-harness) [reply src/services/cd-ng/index.tsx:19092]: @hingerabhinav can you please help here ^^

Comment (hingerabhinav) [reply src/services/cd-ng/index.tsx:19092]: @pankajsingh-harness we are supporting service and env at org/account level now. So artifacts and other contained entities will be using APIs which were earlier designed keeping service at project level in mind. 
Now these params (orgId and projectId) are not required.

**Review by pankajsingh-harness — APPROVED** [review]

**Pull request merged**